split `asm!` parsing and validation

folkertdev opened the pull request:
This PR splits `asm!` parsing and validation into two separate steps.

The parser constructs a `Vec<RawAsmArg>`, with each element corresponding to an argument to one of the `asm!` macros.
The validation then checks things like ordering of arguments or that options are not provided twice.

The motivation is https://github.com/rust-lang/rust/issues/140279, which wants to add `#[cfg(...)]` support to these arguments. This support can now be added in a straightforward way by adding an `attributes: ast::AttrVec` field to `RawAsmArg`. 

An extra reason for this split is that `rustfmt` probably wants to format the assembly at some point (currently that appears to be stubbed out, and the formatting is unstable https://github.com/rust-lang/style-team/issues/152).

r? @ghost (just want to look at CI for now)

cc @ytmimi we discussed asm formatting a little while ago in https://github.com/rust-lang/rustfmt/issues/6526. Am I correct in assuming that `AsmArgs` does not give enough information for formatting, but that `RawAsmArgs` would (it e.g. does not join information from multiple lines). This must have been an issue before?

try-job: aarch64-apple

folkertdev committed:
determine later whether an explicit reg was used

folkertdev committed:
extract operand parser

folkertdev committed:
a new parser generating the exact same error messages

Co-authored-by: Travis Cross <tc@traviscross.com>

folkertdev committed:
delay error for unsupported options

folkertdev committed:
attempt to have rustfmt use the new logic

apparently it doesn't really use the asm parsing at present, so this may work?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14743226924/job/41386114056) [(plain)](https://github.com/rust-lang/rust/commit/3316632a992dae23a8d24700445ee59d067993f0/checks/41386114056/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] toml test:false 0.195
error[E0412]: cannot find type `RawAsmArgs` in this scope
  --> src/tools/rustfmt/src/parse/macros/asm.rs:10:17
   |
10 | ) -> Option<Vec<RawAsmArgs>> {
   |                 ^^^^^^^^^^
   |
  ::: /checkout/compiler/rustc_builtin_macros/src/asm.rs:23:1
   |
23 | pub struct RawAsmArg {
   | -------------------- similarly named struct `RawAsmArg` defined here
   |
help: a struct with a similar name exists
   |
10 - ) -> Option<Vec<RawAsmArgs>> {
10 + ) -> Option<Vec<RawAsmArg>> {
   |
help: you might be missing a type parameter
   |
7  | pub(crate) fn parse_asm<RawAsmArgs>(
   |                        ++++++++++++

error: unused import: `RawAsmArg`
 --> src/tools/rustfmt/src/parse/macros/asm.rs:2:33
  |
2 | use rustc_builtin_macros::asm::{RawAsmArg, parse_raw_asm_args};
  |                                 ^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

```

</details>

folkertdev commented on src/tools/rustfmt/src/parse/macros/asm.rs:
```diff
@@ -1,11 +1,14 @@
 
 #[allow(dead_code)]
-pub(crate) fn parse_asm(context: &RewriteContext<'_>, mac: &ast::MacCall) -> Option<AsmArgs> {
+pub(crate) fn parse_asm(
+    context: &RewriteContext<'_>,
+    mac: &ast::MacCall,
+) -> Option<Vec<RawAsmArg>> {
     let ts = mac.args.tokens.clone();
     let mut parser = super::build_parser(context, ts);
-    parse_asm_args(&mut parser, mac.span(), ast::AsmMacro::Asm).ok()
+    parse_raw_asm_args(&mut parser, mac.span(), ast::AsmMacro::Asm).ok()
 }
```
https://github.com/rust-lang/rustfmt/pull/5191, an old PR for `asm!` formatting, does a bunch of its own parsing that I think would not be needed with `RawAsmArg`.

calebcartwright replied:
Speaking with my t-style hat, we do not yet have any established formatting rules for `asm!`, just a general desire to do so with some fairly concrete thoughts on what those rules would look like.

Speaking with my t-rustfmt hat, this split does look like it'll make our lives much easier to handle `asm!` formatting whenever t-style does finalize the style rules for it, as it does seem to me this would allow us to avoid needing the custom parsing within rustfmt (though Yacin please correct me if I'm missing something)

folkertdev reviewed (commented):
r? @Amanieu (I suspect you're most familiar with this code).

I've left some inline comments to hopefully make reviewing this easier.

folkertdev commented:
https://github.com/rust-lang/rust/pull/140367, specifically the last commit (currently https://github.com/rust-lang/rust/pull/140367/commits/feddf142c09d0481d6b7d38dac5dcdc7194006fa) shows how `#[cfg(...)]` can be built on top of this change.

Amanieu commented on compiler/rustc_builtin_macros/src/asm.rs:
```diff
@@ -18,7 +18,22 @@ use {rustc_ast as ast, rustc_parse_format as parse};
+pub enum RawAsmArgKind {
+    Template(P<ast::Expr>),
+    Operand(Option<Symbol>, ast::InlineAsmOperand),
+    Options(Vec<(Symbol, ast::InlineAsmOptions, Span, Span)>),
```
Can this 4-tuple be replaced with a struct.

folkertdev committed:
rename to get rid of the 'raw' concept

folkertdev committed:
add `AsmOptions` with some named fields

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15089436095/job/42416068538) [(plain)](https://github.com/rust-lang/rust/commit/2e788d15db9ed12d3e179d60164bf34087bc9d9a/checks/42416068538/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] cargo_metadata test:false 0.445
[RUSTC-TIMING] toml_edit test:false 0.927
    Checking toml v0.7.8
[RUSTC-TIMING] toml test:false 0.184
error[E0432]: unresolved imports `rustc_builtin_macros::asm::RawAsmArg`, `rustc_builtin_macros::asm::parse_raw_asm_args`
 --> src/tools/rustfmt/src/parse/macros/asm.rs:2:33
  |
2 | use rustc_builtin_macros::asm::{RawAsmArg, parse_raw_asm_args};
  |                                 ^^^^^^^^^  ^^^^^^^^^^^^^^^^^^ no `parse_raw_asm_args` in `asm`
  |                                 |
  |                                 no `RawAsmArg` in `asm`
  |
help: a similar name exists in the module
  |
2 - use rustc_builtin_macros::asm::{RawAsmArg, parse_raw_asm_args};
2 + use rustc_builtin_macros::asm::{AsmArg, parse_raw_asm_args};
  |
help: a similar name exists in the module
  |
2 - use rustc_builtin_macros::asm::{RawAsmArg, parse_raw_asm_args};
2 + use rustc_builtin_macros::asm::{RawAsmArg, parse_asm_args};
  |

For more information about this error, try `rustc --explain E0432`.
[RUSTC-TIMING] rustfmt_nightly test:false 2.232
error: could not compile `rustfmt-nightly` (lib) due to 1 previous error
```

</details>

Amanieu commented:
r=me once CI passes

bors commented:
:pushpin: Commit d91e3d3e22e071174d9ad28bebc50987dbe19572 has been approved by `Amanieu,traviscross`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu,traviscross d91e3d3e22e071174d9ad28bebc50987dbe19572 -->
<!-- homu: {"type":"Approved","sha":"d91e3d3e22e071174d9ad28bebc50987dbe19572","approver":"Amanieu,traviscross","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

fmease commented:
Failure in rollup: https://github.com/rust-lang/rust/pull/141209#issuecomment-2888861491 (aarch64-apple ).
@bors r-

fmease commented:
Tangential note: Could you please squash some of the micro commits?

folkertdev commented:
Right, of course. I updated the tests based on what I can run locally, I think that covers all of the changes but we should run the try job to make sure. I also squashed/reordered some commits.

@rustbot ready

bors commented:
:hourglass: Trying commit 26e3a5041a3d9bcba04a7b201a47b1b03ccb2b2f with merge 4a6811fe29e2641bcf7a28ff3c652311b32be60e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"26e3a5041a3d9bcba04a7b201a47b1b03ccb2b2f","merge_sha":"4a6811fe29e2641bcf7a28ff3c652311b32be60e"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/15095132608/job/42429976804)
Build commit: 4a6811fe29e2641bcf7a28ff3c652311b32be60e (`4a6811fe29e2641bcf7a28ff3c652311b32be60e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/15095132608/job/42429976804"},"merge_sha":"4a6811fe29e2641bcf7a28ff3c652311b32be60e"} -->

bors commented:
:pushpin: Commit 26e3a5041a3d9bcba04a7b201a47b1b03ccb2b2f has been approved by `Amanieu,traviscross`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu,traviscross 26e3a5041a3d9bcba04a7b201a47b1b03ccb2b2f -->
<!-- homu: {"type":"Approved","sha":"26e3a5041a3d9bcba04a7b201a47b1b03ccb2b2f","approver":"Amanieu,traviscross","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
